docs/reference/src/language/syntax/functions.md

```diff
@@ -14,27 +35,158 @@ Functions can be annotated with the `pure` keyword.
+
+When called with an element name (or `self`, `parent` or `root`), the function must be defined on that
+element. Name resolution does not look at ancestor elements in this case. Note that this means
+calling a function without an element name is _not_ equivalent to calling it with `self`.
+
+Multiple functions with the same name are allowed in the same component, as long as they are defined
+on different elements. Therefore it is possible for a function to shadow another function from an ancestor
+element.
+
+```slint,no-preview
+export component Example {
+    property <int> secret_number: my-function();
+    public pure function my-function() -> int {
+        return 1;
+    }
+
+    VerticalLayout {
+        public pure function my-function() -> int {
+            return 2;
+        }
+
+        Text {
+            text: "The secret number is " + my-function();
+            public pure function my-function() -> int {
+                return 3;
+            }
+        }
+
+        Text {
+            text: "The other secret number is " + my-function();
+        }
+    }
+}
+```
+
+In the example above, the property `secret_number` will be set to 1, and the text labels will say "The
+secret number is 3" and "The other secret number is 2".
+
+## Function visibility
+
+By default, functions are private and cannot be accessed from other components.
+
 However, their accessibility can be modified using the `public` or `protected` keywords.
 
-- A function annotated with `public` can be accessed by any component.
+- A root-level function annotated with `public` can be accessed by any component.
+
+To access such a function from a different component, you always need a target, which in practice
+means the calling component must declare the called component as one of its child elements.
+
+```slint,no-preview
+export component HasFunction {
+    public pure function double(x: int) -> int {
+        return x * 2;
+    }
+}
+
+export component CallsFunction {
+    property <int> test: my-friend.double(1);
+
+    my-friend := HasFunction {
+    }
+}
+```
+
+If a function is declared in a child element, even if marked public, it is not possible to call it
+from another component, as the child elements themselves are not public and a valid target for the
+function does not exist:
+
+```slint,no-preview
+export component HasFunction {
+    t := Text {
+        public pure function double(x: int) -> int {
+            return x * 2;
+        }
+    }
+}
+
+export component CallsFunction {
+    // Compiler error!
+    // property <int> test: my-friend.t.double(1);
+
+    my-friend := HasFunction {
+    }
+}
+```
+
 - A function annotated with `protected` can only be accessed by components that directly inherit from it.
 
-## Example
+Functions, even marked public, cannot be exported and cannot be called from backend code (in Rust, C++,
+JS, etc.). Declare a [callback](callbacks.md) which can call the function.
```

Comment:
Function can be called from backend code. (invoke_...)
The difference with callback is that they cannot be replaced with a callback handler.

Reply:
Also, your point about the difference between this and callbacks made me take a stab at a "functions vs callbacks" section, could you look over it and let me know if I got anything wrong?

Reply:
Indeed, falling function from native code is missing from that documentation 🙈